### PR TITLE
onChange (newValue, oldValue, keyPath)

### PR DESCRIPTION
**Pull request by neftaly**
This PR mirrors the API in [contrib/cursor/index.js#L42](https://github.com/facebook/immutable-js/blob/master/contrib/cursor/index.d.ts#L42), by swapping the args in the updater callback, and adding `keyPath`.
```js
onChange?: (newValue: any, oldValue?: any, keyPath?: Array<any>) => any
```

### CODE REVIEW

**Comment by neftaly** [comment]
Note that this isn't to be merged - just getting a feel for your opinion on it.

**Comment by neftaly** [comment]
Arg order in updates needs to be swapped in immutable-cursor for this to work (`nextValue, prevValue` not `prevValue, nextValue`)
```js
// Example from immutable-cursor/README.md
const cursor = Cursor.from(data, (prevValue, nextValue) => {
  console.log(newData)
  // First call: $> Map { "a": 2, "b": 2 }
  // Second call: $> Map { "a": 2, "b": 3 }
});

// Advantage is that if you don't care about prevValue, you don't need to declare it
const cursor = Cursor.from(data, nextValue => {
   ...
});
```

this would be accomplished by:
```js
const inMemory = (data, transition) => {
  let rootState = data;

  return {
    write(fn, ...context) {
      const oldState = this.read();
      const newState = fn(oldState);
      transition(newState, oldState, ...context);
      rootState = newState;
      return rootState;
    },
    read() {
      return rootState;
    }
  };
};

```

**Comment by neftaly** [comment]
https://github.com/facebook/immutable-js/blob/master/contrib/cursor/index.d.ts#L40

```js
export function from(
  collection: Immutable.Collection<any, any>,
  onChange?: (newValue: any, oldValue?: any, keyPath?: Array<any>) => any
): Cursor;
export function from(
  collection: Immutable.Collection<any, any>,
  keyPath: Array<any>,
  onChange?: (newValue: any, oldValue?: any, keyPath?: Array<any>) => any
): Cursor;
export function from(
  collection: Immutable.Collection<any, any>,
  key: any,
  onChange?: (newValue: any, oldValue?: any, keyPath?: Array<any>) => any
): Cursor;
```

File: package.json [hunk 1]
```diff
@@ -21,7 +21,7 @@
   },
   "homepage": "https://github.com/redbadger/immutable-cursor",
   "dependencies": {
-    "atom-store": "0.0.2",
+    "atom-store": "jameshopkins/atom-store",
```
Comment (neftaly) [inline package.json]: Temporary (atom-store update needs to be published to NPM)

File: src/utils.js [hunk 2]
```diff
@@ -78,14 +78,19 @@ export function subCursor(cursor, keyPath, value) {
   );
 }
 
-export function updateCursor(cursor, changeFn) {
+export function updateCursor(cursor, changeFn, keyPath) {
```
Comment (neftaly) [inline src/utils.js]: I meant to make this a separate commit

File: package.json [hunk 3]
```diff
@@ -21,8 +21,8 @@
   },
   "homepage": "https://github.com/redbadger/immutable-cursor",
   "dependencies": {
-    "atom-store": "0.0.2",
-    "immutable": "3.7.6"
+    "atom-store": "0.0.3",
```
Comment (jameshopkins) [inline package.json]: We'll need to cut a new major version of `atom-store` due to https://github.com/jameshopkins/atom-store/commit/e809f638eb35566a38e72e4a5d8fb02630512ebd and potentially any other changes to the interface we want to rollup into the next release.

File: package.json [hunk 4]
```diff
@@ -21,8 +21,8 @@
...
+    "immutable": "^3.7.6"
```
Comment (jameshopkins) [inline package.json:25]: Nice catch. I missed that in https://github.com/redbadger/immutable-cursor/pull/13/commits/3b40e2e78e9e022e6829ba89671ef65c45639c59

File: README.md [hunk 5]
```diff
@@ -17,7 +17,7 @@ A Clojure-inspired atom is placed above the cursor composition, and is the only
 ```js
 const data = Immutable.fromJS({a: 1, b: 2});
 
-const cursor = Cursor.from(data, (prevValue, nextValue) => {
+const cursor = Cursor.from(data, (nextValue, prevValue, keyPath) => {
   console.log(newData)
```
Comment (jameshopkins) [inline README.md]: Due to my own stupidity, we're `console.log`'ing an undeclared variable here. Are you able to sneak in a new commit whilst you're touching this file, to update this to `nextValue`?

I'm actually thinking whether we should also include `prevValue` and `keyPath` in the `console.log` too for explicitness. What do you think?

**Comment by jameshopkins** [comment]
Awesome 👍. The original aim of this project was to exactly mirror the [`contrib/cursor`](https://github.com/facebook/immutable-js/tree/master/contrib/cursor) API. Not sure how the disparity with the argument order crept in!

**Comment by jameshopkins** [comment]
We should have at least one test to ensure that the correct `keyPath` is being passed as the third argument to the `onChange` callback.

File: src/cursor.js [hunk 6]
```diff
@@ -10,8 +10,21 @@
 import {valToKeyPath, makeCursor} from './utils';
 import createAtom from 'atom-store';
 
+const inMemory = (initial, transition) => {
+  let rootState = initial;
+  const read = () => rootState;
+  const write = (fn, ...context) => {
+    const oldState = read();
+    const newState = fn(oldState);
+    transition(newState, oldState, ...context);
```
Comment (jameshopkins) [inline src/cursor.js]: I'm in two minds as to whether providing the ability to add arbitrary arguments from `write()` through to the `onChange` callback(s), should be the responsibility of the consumer application, or in fact, should be an implementation concern for the core `atom-store` plugin itself. My hunch, is that it _could_ be a relatively common requirement from consumers other than just `immutable-cursor`.

What are your thoughts?

Comment (neftaly) [reply src/cursor.js]: Agreed :) After https://github.com/jameshopkins/atom-store/pull/2 is merged I plan to make a atom-store PR for both `context` and the arg order swap

File: README.md (hunk 5 above)
Comment (neftaly) [reply README.md]: Sure

File: src/cursor.js (hunk 6 above)
Comment (jameshopkins) [reply src/cursor.js]: Great. I'd be keen to wait for this to go into `atom-store`, so that `immutable-cursor` can rely on the revised core `inMemory` plugin, instead.

File: README.md [hunk 7]
```diff
@@ -17,19 +17,18 @@ A Clojure-inspired atom is placed above the cursor composition, and is the only
...
-  console.log(newData)
-  // First call: $> Map { "a": 2, "b": 2 }
-  // Second call: $> Map { "a": 2, "b": 3 }
+const cursor = Cursor.from(data, (nextValue, prevValue, keyPath) => {
+  console.log('Value changed from', prevValue, 'to', nextValue, 'at', keyPath);
 });
 
 // Multiple update transactions are serialised.
 cursor.set('a', 2);
+// Value changed from Map { a: 1, b: 2 } to Map { a: 2, b: 2 } at []
```
Comment (neftaly) [inline README.md]: `{ a: 2, b: 2 } at []` or `2 at ['a']`?

File: README.md [hunk 8]
```diff
@@ -17,19 +17,18 @@ A Clojure-inspired atom is placed above the cursor composition, and is the only
...
 cursor.set('b', 3);
+// Value changed from Map { a: 2, b: 2 } to Map { a: 2, b: 3 } at []
```
Comment (neftaly) [inline README.md]: `{ a: 2, b: 3 } at []` or `3 at ['b']`?

Comment (jameshopkins) [reply README.md]: I think the former is preferable, since it's good to highlight the update serialisation of multiple transactions

File: README.md (hunk 7 above)
Comment (jameshopkins) [reply README.md]: See https://github.com/redbadger/immutable-cursor/pull/15/files/b53d43c0c9e8dc7c900eaee8775c8c638695504d#r112674508

File: package.json (hunk 3 above)
Comment (jameshopkins) [inline package.json]: Am happy to publish a new release of `atom-store` at `0.0.3`, since the [change in argument order](https://github.com/jameshopkins/atom-store/commit/e809f638eb35566a38e72e4a5d8fb02630512ebd) has since been [reverted](https://github.com/jameshopkins/atom-store/commit/f9728c86eb7002b806b2f380459c4ec8d92bad1e).

**Comment by jameshopkins** [comment]
All that is left now, is [at least one test](https://github.com/redbadger/immutable-cursor/pull/15#issuecomment-295775653)

File: package.json (hunk 3 above)
Comment (jameshopkins) [inline package.json:24]: We need to upgrade the `atom-store` dependency through Yarn too.

**Comment by neftaly** [comment]
Done!

**Comment by jameshopkins** [comment]
Are you able to upgrade atom-store through Yarn too? When this had been done, I'll merge and cut a new release

**Comment by neftaly** [comment]
Sure - immutable is auto-updated in yarn too, is that ok?

**Comment by jameshopkins** [comment]
I think an update of `immutable` too should be out of scope of this PR - although we definitely should create an issue tracking the progress of v4, anyway.

For the time being, I think `yarn upgrade atom-store` should work.

**Comment by neftaly** [comment]
Ok, fixed - but just a warning that yarn is going to do this anyway every time you run `yarn` (as it's `yarn.lock` not `package.json`).

**Comment by neftaly** [comment]
Can you please publish (2.0.0) after merging, btw? I'd like to make a PR for immstruct.

**Comment by jameshopkins** [comment]
> ... a warning that yarn is going to do this anyway every time you run ’yarn’

Not _entirely_ sure I understand what you're alluding to.`yarn` isn't a mutative action in this context, and will respect the versions defined in `yarn.lock`.

I will publish 2.0.0 at the next available opportunity, which'll be Sunday.
